runtime/vm/FastJNI_com_ibm_crypto_plus_provider_icc_NativeInterface.cpp

```diff
@@ -0,0 +1,415 @@
+			j9sl_lookup_name(*cryptoLibrary, (char*)encryptUpdate_name, (UDATA*)&(cryptFuncs->CIPHER_encryptUpdate), NULL);
+			j9sl_lookup_name(*cryptoLibrary, (char*)decryptUpdate_name, (UDATA*)&(cryptFuncs->CIPHER_decryptUpdate), NULL);
+			j9sl_lookup_name(*cryptoLibrary, (char*)zKMC_name, (UDATA*)&z_kmc_native, NULL);
+		
+#ifdef CRYPTO_DEBUG
+			printf("** CIPHER_encryptFinal: %p\n",CIPHER_encryptFinal);
+			printf("** CIPHER_decryptFinal: %p\n",CIPHER_decryptFinal);
+			printf("** CIPHER_encryptUpdate: %p\n",cryptFuncs->CIPHER_encryptUpdate);
+			printf("** CIPHER_decryptUpdate: %p\n",cryptFuncs->CIPHER_decryptUpdate);
+			printf("** z_kmc_native: %p\n",z_kmc_native);
+#endif
+		
+			cryptFuncs->CIPHER_encryptFinalPtr.CIPHER_op = CIPHER_encryptFinal;
+			cryptFuncs->CIPHER_encryptUpdatePtr.CIPHER_op = cryptFuncs->CIPHER_encryptUpdate;
+			cryptFuncs->CIPHER_decryptFinalPtr.CIPHER_op = CIPHER_decryptFinal;
+			cryptFuncs->CIPHER_decryptUpdatePtr.CIPHER_op = cryptFuncs->CIPHER_decryptUpdate;
+			cryptFuncs->z_kmc_nativePtr.z_kmc_native = z_kmc_native;
+	
+			currentThread->javaVM->jniCryptoFunctions = (void*)cryptFuncs; 
+		} 
+		
+		omrthread_monitor_exit(mutex);
+	}
+}
+
+static VMINLINE bool
+isContiguousRange2(J9VMThread *currentThread, UDATA index, UDATA count) {
+	UDATA arrayletLeafSizeInElements = currentThread->javaVM->arrayletLeafSize;
+
+	UDATA lastIndex = index;
+	lastIndex += count;
+	lastIndex -= 1;
+
+	lastIndex ^= index;
+
+	return lastIndex < arrayletLeafSizeInElements;
+}
+
+/* This routine handles encrypt and decrypt operations for AES operations.
+ * This routines calls the passed function pointer on the input data, this function may encrypt or decrypt.
+ * The input and output are always byte arrays.
+ *
+ * If the input data is very large and VM Access may be acquired for too extended a duration. In that case, 
+ * VMAccess is released and re-acquired after a fixed number of bytes have been processed to allow the GC to operate.
+ *
+ * The maxinum number of bytes to be processed at a single operation was selected as follows: vmaccess should not be acquired for no more than 0.1ms. 
+ * Benchmarks indicate current performance at 2020 is at 620 MB/s for AES-CBC encrypt at 16 threads on x86 linux.
+ * Other scenarios & platforms are all faster. Using 620 MB/s as worst case performance, we have 62KB processed at 0.1ms
+ */
+
+#define OP_SIZE_LIMIT 63488
+
+VMINLINE jint AESLaunch(J9VMThread *currentThread, jlong iccContextId, jlong iccCipherId, j9object_t input, jint inputOffset,jint inputLen, j9object_t output, jint outputOffset, jboolean needsReinit, CIPHER_t* cbcOpUpdate, cryptoFunc_t cbcOp, jlong paramPointer, jint mode) {
+
+	j9object_t in = input;
+	j9object_t out = output;
+	int inputIndex = inputOffset; 
+	int outputIndex = outputOffset;
+	void* inputAddress = NULL;
+	void* outputAddress = J9JAVAARRAY_EA(currentThread, out, outputIndex, U_8);
+
+	jint totalBytes = inputLen;
+
+	if (totalBytes > 0) {
```

Comment:
No need for this check.

Reply:
if the value is negative, no need to handle it in vm code - the JCL code will take care of throwing exceptions etc as needed. negative value doesn't impact the functionality of the vm code - a single shot call will be invoked with negative bytes and then return.

Reply:
OK, then the check should be removed (and comment this fact in the function doc).

Reply:
If totalBytes is zero, the JCL is indicating that the input array is null and therefore I skip the line and leave inputAddress at NULL. 
``inputAddress = J9JAVAARRAY_EA(currentThread, in, inputIndex, U_8);``

Reply:
Given that:

- the length will essentially never be 0
- the array macro will work fine with a 0-sized array
- 0-sized arrays are handled by the subsequent code

The check is not needed.

Reply:
It's still worth commenting on the ensured validity of the input arguments.

Reply:
will the macro work if ``in`` is a null in java?

Reply:
No, it will not work with NULL. Perhaps check for that instead of the size, assuming that both will be 0 for the size computation case.